mem_domain: add arch specfic implementation for k_mem_domain_remove_thread and add_thread

AdithyaBaglody opened the pull request:
If the thread id is same as current then handle the cleanup of the
memory domain.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: mem_domain: Add arch specfic destroy for remove thread API.

If the thread id is same as current then handle the cleanup of the
memory domain.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

SavinayDharmappa reviewed (commented):
is the use of local ptr "domain" is necessary?

chunlinhan commented on kernel/mem_domain.c:
```diff
@@ -164,13 +164,18 @@ void k_mem_domain_add_thread(struct k_mem_domain *domain, k_tid_t thread)
+	}
 
 	sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
-	thread->mem_domain_info.mem_domain = NULL;
```
thread->mem_domain_info.mem_domain needs to be set to NULL for arch arm. (so does x86, I think.)
It seems the local variable 'domain' is unnecessary.

AdithyaBaglody replied:
The local variable is manual optimization. I feel instead of dereferencing it multiple times a single dereference should be better.
And coming to the functional part I will still set the mem_domain structure to NULL.

chunlinhan replied:
Even you set domain = NULL, the thread->mem_domain_info.mem_domain is still keep the same value.
Or anything I misunderstand?

andrewboie replied:
> The local variable is manual optimization. I feel instead of dereferencing it multiple times a single dereference should be better.

Are you trying to out-smart the compiler?
At any rate, this looks wrong, pointer to the mem domain in the thread struct is NOT being cleared. Did you think that "domain = NULL" would somehow do this?

AdithyaBaglody replied:
oh my mistake I thought I did a double pointer here. will revert this line back.

andrewboie commented on kernel/mem_domain.c:
```diff
@@ -164,13 +164,18 @@ void k_mem_domain_add_thread(struct k_mem_domain *domain, k_tid_t thread)
 	key = irq_lock();
+	domain = thread->mem_domain_info.mem_domain;
+	if (_current == thread) {
+		_arch_mem_domain_destroy(domain);
```
What happens if other threads were members of the domain?

AdithyaBaglody replied:
Other threads doesn't matter since this will just set the page properties for the current thread. This arch specific destroy doesn't manipulate the mem_domain struct. So for the other threads which are a part of this mem domain will be configured when they get swapped in.

AdithyaBaglody committed:
kernel: mem_domain: Add an arch interface to configure memory domain

Add an architecure specfic code for the memory domain
configuration. This is needed to support a memory domain API
k_mem_domain_add_thread.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: mem_domain: Add to current thread should configure immediately.

when a current thread is added to a memory domain the pages/sections
must be configured immediately.
A problem occurs when we add a thread to current and then drop
down to usermode. In such a case memory domain will become active
the next time a swap occurs.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
Delta:
1. added another patch for k_mem_domain_add_thread
2. involves changing the arch.h
3. x86 swap now calls the arch specific api to configure mem domains

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -262,12 +262,18 @@ static inline void _x86_mem_domain_pages_update(struct k_mem_domain *mem_domain,
+/* Load the partitions of the thread. */
+void _arch_mem_domain_configure(struct k_thread *thread)
+{
+	_x86_mmu_mem_domain_load(thread);
```
why does _x86_mmu_mem_domain_load() need to be a separate function?
Just move what it does here, and delete it.

AdithyaBaglody commented:
Delta:
Fixed Andrew's review comments.
